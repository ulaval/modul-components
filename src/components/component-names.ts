export const ACCORDION_GROUP_NAME: string = 'm-accordion-group';
export const ACCORDION_NAME: string = 'm-accordion';
export const LIST_ITEM_NAME: string = 'm-list-item';
export const BUTTON_NAME: string = 'm-button';
export const BUTTON_GROUP_NAME: string = 'm-button-group';
export const CAROUSEL_NAME: string = 'm-carousel';
export const CAROUSEL_ITEM_NAME: string = 'm-carousel-item';
export const CHECKBOX_NAME: string = 'm-checkbox';
export const DATEFIELDS_NAME: string = 'm-datefields';
export const DATEPICKER_NAME: string = 'm-datepicker';
export const DIALOG_NAME: string = 'm-dialog';
export const DROPDOWN_NAME: string = 'm-dropdown';
export const DROPDOWN_ITEM_NAME: string = 'm-dropdown-item';
export const DROPDOWN_GROUP_NAME: string = 'm-dropdown-group';
export const DYNAMIC_TEMPLATE_NAME: string = 'm-dynamic-template';
export const FILE_SELECT_NAME: string = 'm-file-select';
export const FILE_UPLOAD_NAME: string = 'm-file-upload';
export const EDIT_WINDOW_NAME: string = 'm-edit-window';
export const FLEX_TEMPLATE_NAME: string = 'm-flex-template';
export const I18N_NAME: string = 'm-i18n';
export const ICON_NAME: string = 'm-icon';
export const ICON_BUTTON_NAME: string = 'm-icon-button';
export const INPUT_STYLE_NAME: string = 'm-input-style';
export const LIMIT_TEXT_NAME: string = 'm-limit-text';
export const LINK_NAME: string = 'm-link';
export const MESSAGE_NAME: string = 'm-message';
export const MODAL_NAME: string = 'm-modal';
export const NAVBAR_NAME: string = 'm-navbar';
export const NAVBAR_ITEM_NAME: string = 'm-navbar-item';
export const MENU_NAME: string = 'm-menu';
export const MENU_ITEM_NAME: string = 'm-menu-item';
export const PANEL_NAME: string = 'm-panel';
export const PHONE_NUMBER_NAME: string = 'm-phone-number';
export const POPPER_NAME: string = 'm-popper';
export const POPUP_NAME: string = 'm-popup';
export const PROGRESS_NAME: string = 'm-progress';
export const RADIO_NAME: string = 'm-radio';
export const RADIO_GROUP_NAME: string = 'm-radio-group';
export const RADIO_STYLE_NAME: string = 'm-radio-style';
export const SCROLL_TOP_NAME: string = 'm-scroll-top';
export const SIDEBAR_NAME: string = 'm-sidebar';
export const SLIDER_NAME: string = 'm-slider';
export const SPINNER_NAME: string = 'm-spinner';
export const STATUS_NAME: string = 'm-status';
export const STEP_NAME: string = 'm-step';
export const STEPPERS_NAME: string = 'm-steppers';
export const STEPPERS_ITEM_NAME: string = 'm-steppers-item';
export const SWITCH_NAME: string = 'm-switch';
export const TAB_PANEL_NAME: string = 'm-tab-panel';
export const TABS_NAME: string = 'm-tabs';
export const TEMPLATE_NAME: string = 'm-template';
export const TEXTAREA_NAME: string = 'm-textarea';
export const TEXTFIELD_NAME: string = 'm-textfield';
export const TIMEPICKER_NAME: string = 'm-timepicker';
export const TOOLTIP_NAME: string = 'm-tooltip';
export const UPLOAD_NAME: string = 'm-upload';
export const UPLOAD_INPUT_NAME: string = 'm-upload-input';
export const UPLOAD_DRAGDROP_NAME: string = 'm-upload-dragdrop';
export const UPLOAD_FILESLIST_NAME: string = 'm-upload-fileslist';
export const VALIDATION_MESSAGE_NAME: string = 'm-validation-message';
