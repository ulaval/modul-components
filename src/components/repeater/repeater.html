<div class="m-repeater">

    <component :is="tag"
               v-if="list.length">
        <template v-for="(item, index) in list">

            <slot :props="getRowProps(item, index)"
                  :listeners="getRowListeners(item, index)"
                  name="row">

                <component :is="'li'"
                           class="m-repeater__row"
                           :key="getItemKey(item, index)">

                    <div class="m-repeater__col">
                        <slot :props="getItemProps(item, index)"
                              name="item"></slot>
                    </div>

                    <m-icon-button v-if="canDelete"
                                   class="m-repeater__delete"
                                   :ref="'deleteBtn' + index"
                                   @click="onDeleteBtnClick(index)"
                                   icon-name="m-svg__delete">{{ deleteBtnLabel }}</m-icon-button>
                </component>
            </slot>
        </template>
    </component>

    <div v-else-if="emptyMessage"
         class="m-repeater__empty-row">
        {{ emptyMessage }}
    </div>

    <!-- TODO: Replace with add component when it becomes available.  -->
    <m-link class="m-repeater__add"
            v-if="canAdd"
            mode="button"
            icon-name="m-svg__add-circle-filled"
            icon-size="24px"
            ref="addBtn"
            @click="onAddBtnClick()">
        {{ addBtnLabel }}
    </m-link>
</div>
