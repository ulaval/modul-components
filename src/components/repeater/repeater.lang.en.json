{
    "m-repeater": {
        "add": "Add",
        "delete": "Delete"
    }
}
