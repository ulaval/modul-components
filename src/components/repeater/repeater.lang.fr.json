{
    "m-repeater": {
        "add": "Ajouter",
        "delete": "Supprimer"
    }
}
