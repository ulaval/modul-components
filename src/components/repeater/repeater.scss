@import 'commons';

$m-repeater-translate-offset: 50px;
$m-repeater-spacing: $m-spacing;

.m-repeater {
    ul {
        padding: 0;
        margin: 0;
    }

    &__row {
        display: flex;
        align-items: center;
        background-color: $m-color--white;
        position: relative;
        z-index: 3;

        &:not(.m--is-empty) {
            margin-bottom: $m-repeater-spacing;
        }

        &.m--is-empty {
            border: 1px solid transparent;
            padding: $m-repeater-spacing;
            z-index: 1;
        }
    }

    &__col {
        &:first-child {
            border: 1px solid $m-color--border;
            padding: $m-repeater-spacing;
            flex: 1;
        }
    }

    &__footer {
        display: block;
        position: relative;
        z-index: 2;
        background-color: $m-color--white;

        // fix wierd transition-group behavior.
        &.m--is-slot-row.m--is-slot-row {
            transition: none;
            transform: none;
        }
    }

    .m--is-enter-active,
    .m--is-leave-active,
    .m--is-move {
        transition: 0.5s cubic-bezier(0.59, 0.12, 0.34, 0.95);
        transition-property: opacity, transform;
    }

    .m--is-enter {
        opacity: 0;
        transform: translateX($m-repeater-translate-offset) scaleY(0.5);

        &.m--is-empty {
            transform: translateY($m-repeater-translate-offset);
        }
    }

    .m--is-enter-to {
        opacity: 1;
        transform: translateX(0) scaleY(1);
    }

    .m--is-leave-active {
        position: absolute;
        width: 100%;
    }

    .m--is-leave-to {
        opacity: 0;
        transform: scaleY(0);
        transform-origin: center top;

        &.m--is-empty {
            transform: translateY($m-repeater-translate-offset);
        }
    }
}
