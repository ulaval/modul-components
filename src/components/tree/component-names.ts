export const TREE_NODE_NAME: string = 'm-tree-node';
export const TREE_ICON_NAME: string = 'm-tree-icon';
