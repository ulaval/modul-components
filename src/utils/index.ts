import Vue, { PluginObject } from 'vue';

import MediaQueriesPlugin from './media-queries/media-queries';
import ModulPlugin from './modul/modul';
import MessagesPlugin from './i18n/i18n';
import HttpPlugin, { HttpPluginOptions } from './http/http';
import SpritesPlugin from './svg/sprites';
import * as TouchPlugin from 'vue-touch';
import ConfirmPlugin from './modal/confirm';
import AlertPlugin from './modal/alert';
import FilePlugin from '../utils/file/file';
import { PortalPluginInstall } from 'portal-vue';
import LoggerPlugin, { ConsoleOptions } from './logger/logger';

export interface UtilsPluginOptions {
    httpPluginOptions?: HttpPluginOptions;
    consoleOptions?: ConsoleOptions;
}

const UtilsPlugin: PluginObject<any> = {
    install(v, options): void {
        if (!v.prototype.$log) {
            Vue.use(LoggerPlugin, options ? options.consoleOptions : undefined);
        } else if (options) {
            v.prototype.$log.setConsoleOptions(options.consoleOptions);
        }

        Vue.use(MediaQueriesPlugin);
        Vue.use(ModulPlugin);
        Vue.use(MessagesPlugin);
        Vue.use(HttpPlugin, options ? options.httpPluginOptions : undefined);
        Vue.use({ install: PortalPluginInstall });
        Vue.use(SpritesPlugin);
        Vue.use(TouchPlugin, { name: 'v-touch' });
        Vue.use(ConfirmPlugin);
        Vue.use(AlertPlugin);
        Vue.use(FilePlugin);
    }
};

export default UtilsPlugin;
