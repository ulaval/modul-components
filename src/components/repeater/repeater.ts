import { PluginObject } from 'vue';
import { Component, Emit, Prop } from 'vue-property-decorator';
import { ModulVue } from '../../utils/vue/vue';
import AddPlugin from '../add/add';
import { REPEATER_NAME } from '../component-names';
import IconButtonPlugin from '../icon-button/icon-button';
import WithRender from './repeater.html?style=./repeater.scss';

export interface MRepeaterItem { }

export interface MRepeaterOperations {
    canAdd: boolean;
    canDelete: boolean;
}

export interface MRepeaterItemProps<T = MRepeaterItem> {
    item: T;
    index: number;
}

export interface MRepeaterRowProps<T = MRepeaterItem> extends MRepeaterItemProps<T> {
    canDelete: boolean;
}

export interface MRepeaterRowListeners {
    onDelete(): void;
}

@WithRender
@Component
export class MRepeater extends ModulVue {
    @Prop({
        required: true
    })
    list: MRepeaterItem[];

    @Prop()
    addButtonLabel?: string;

    @Prop()
    deleteButtonLabel?: string;

    @Prop()
    emptyMessage?: string;

    @Prop({
        default: 'ul'
    })
    tag: string;

    @Prop()
    itemKey?: string;

    @Prop({
        default: (): MRepeaterOperations => ({
            canAdd: true,
            canDelete: true
        })
    })
    operations: MRepeaterOperations;

    @Prop({
        default: 0
    })
    minItemCount: number;

    @Prop({
        default: Infinity
    })
    maxItemCount: number;

    mounted(): void {
        if (!this.$scopedSlots.row && !this.$scopedSlots.item) {
            throw new Error('MRepeater requires content to be provided through row or item slot.');
        }

        // tslint:disable-next-line:no-console
        console.log(this.$scopedSlots.row);

    }

    @Emit('add')
    onAddBtnClick(): void { }

    @Emit('delete')
    onDeleteBtnClick(index: number): void { }

    getRowProps(item: MRepeaterItem, index: number): MRepeaterRowProps {
        return {
            ...this.getItemProps(item, index),
            canDelete: this.canDelete
        };
    }

    getRowListeners(item: MRepeaterItem, index: number): MRepeaterRowListeners {
        return {
            onDelete: () => this.onDeleteBtnClick(index)
        };
    }

    getItemKey(item: MRepeaterItem, index: number): unknown | number {
        return this.itemKey ? item[this.itemKey] : index;
    }

    getItemProps(item: MRepeaterItem, index: number): MRepeaterItemProps {
        return {
            item, index
        };
    }

    get addBtnLabel(): string {
        return this.addButtonLabel || this.$i18n.translate('m-repeater:add');
    }

    get deleteBtnLabel(): string {
        return this.deleteButtonLabel || this.$i18n.translate('m-repeater:delete');
    }

    get canAdd(): boolean {
        return this.operations.canAdd && this.list.length < this.maxItemCount;
    }

    get canDelete(): boolean {
        return this.operations.canDelete && this.list.length > this.minItemCount;
    }

    get itemTag(): string {
        return this.tag === 'ul' ? 'li' : 'span';
    }

    get hasRowSlot(): boolean {
        return this.$scopedSlots.row ? true : false;
    }
}

const RepeaterPlugin: PluginObject<any> = {
    install(v): void {
        v.use(AddPlugin);
        v.use(IconButtonPlugin);
        v.component(REPEATER_NAME, MRepeater);
    }
};

export default RepeaterPlugin;

