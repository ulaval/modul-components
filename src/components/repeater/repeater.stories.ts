import { withA11y } from '@storybook/addon-a11y';
import { number, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/vue';
import { componentsHierarchyRootSeparator } from '../../../conf/storybook/utils';
import uuid from '../../utils/uuid/uuid';
import { REPEATER_NAME } from '../component-names';
import { MRepeater } from './repeater';

storiesOf(`${componentsHierarchyRootSeparator}${REPEATER_NAME}`, module)
    .addDecorator(withA11y)
    .addDecorator(withKnobs)
    .add('Empty', () => ({
        components: { MRepeater },
        template: `
            <${REPEATER_NAME}
                :empty-message="'The list is empty'"
                :list="list"
                :operations="operations">
                <template slot="item" slot-scope="{}">
                </template>
            </${REPEATER_NAME}>
        `,
        data: () => ({
            list: [],
            operations: {
                canAdd: false,
                cadDelete: false
            }
        }),
        methods: {
            onAdd(): void {
                (this as any).list.push({ id: uuid.generate() });
            },
            onDelete(index): void {
                (this as any).list.splice(index, 1);
            }
        }
    }));

storiesOf(`${componentsHierarchyRootSeparator}${REPEATER_NAME}`, module)
    .addDecorator(withA11y)
    .addDecorator(withKnobs)
    .add('With items (Custom item template)', () => ({
        components: { MRepeater },
        template: `
            <${REPEATER_NAME}
                :item-key="'id'"
                :empty-message="'The list is empty'"
                :min-item-count="minItemCount"
                :max-item-count="maxItemCount"
                :list="list"
                @add="onAdd"
                @delete="onDelete">
                <template slot="item" slot-scope="{ props }">
                    Item # {{ props.index }} {{ props.item.id }}
                </template>
            </${REPEATER_NAME}>
        `,
        data: () => ({
            list: [{ id: uuid.generate() }, { id: uuid.generate() }]
        }),
        props: {
            minItemCount: {
                default: number('minItemCount', 0)
            },
            maxItemCount: {
                default: number('maxItemCount', Infinity)
            }
        },
        methods: {
            onAdd(): void {
                (this as any).list.push({ id: uuid.generate() });
            },
            onDelete(index): void {
                (this as any).list.splice(index, 1);
            }
        }
    }));

storiesOf(`${componentsHierarchyRootSeparator}${REPEATER_NAME}`, module)
    .addDecorator(withA11y)
    .addDecorator(withKnobs)
    .add('With items (Custom row template)', () => ({
        components: { MRepeater },
        template: `
            <${REPEATER_NAME}
                :min-item-count="minItemCount"
                :max-item-count="maxItemCount"
                :list="list"
                @add="onAdd"
                @delete="onDelete">
                <li slot="row" slot-scope="{ props, listeners }" :key="props.item.id">
                    <button @click="listeners.onDelete" :disabled="!props.canDelete">Delete</button>
                    Item # {{ props.index }} {{ props.item.id }}
                </li>
            </${REPEATER_NAME}>
        `,
        data: () => ({
            list: [{ id: uuid.generate() }, { id: uuid.generate() }]
        }),
        props: {
            minItemCount: {
                default: number('minItemCount', 0)
            },
            maxItemCount: {
                default: number('maxItemCount', Infinity)
            }
        },
        methods: {
            onAdd(): void {
                (this as any).list.push({ id: uuid.generate() });
            },
            onDelete(index): void {
                (this as any).list.splice(index, 1);
            }
        }
    }));

storiesOf(`${componentsHierarchyRootSeparator}${REPEATER_NAME}`, module)
    .addDecorator(withA11y)
    .addDecorator(withKnobs)
    .add('Operations', () => ({
        components: { MRepeater },
        template: `
            <${REPEATER_NAME}
                :item-key="'id'"
                :list="list"
                :operations="operations">
                <template slot="item" slot-scope="{ props }">
                    #{{ props.index }} Custom template for item (id: {{ props.item.id }} name: {{ props.item.name }})
                </template>
            </${REPEATER_NAME}>
        `,
        data: () => ({
            list: [{ id: 324525, name: 'Albert Einstein' }, { id: 64564556, name: 'Nikola Tesla' }],
            operations: {
                canAdd: false,
                cadDelete: false
            }
        })
    }));
