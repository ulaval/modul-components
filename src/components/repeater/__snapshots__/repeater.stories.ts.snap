// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots components|m-repeater Empty 1`] = `
<div
  class="m-repeater"
>
  <ul
    class="m-repeater__list"
  >
    <li
      class="m-repeater__row m--is-empty"
    >
      
            The list is empty
        
    </li>
    <li
      class="m-repeater__footer"
    >
      <!---->
    </li>
  </ul>
</div>
`;

exports[`Storyshots components|m-repeater Operations 1`] = `
<div
  class="m-repeater"
>
  <ul
    class="m-repeater__list"
  >
    <li
      class="m-repeater__row"
    >
      <div
        class="m-repeater__col"
      >
        
                    #0 Custom template for item (id: 324525 name: Albert Einstein)
                
      </div>
       
      <!---->
    </li>
    <li
      class="m-repeater__row"
    >
      <div
        class="m-repeater__col"
      >
        
                    #1 Custom template for item (id: 64564556 name: Nikola Tesla)
                
      </div>
       
      <!---->
    </li>
    <li
      class="m-repeater__footer"
    >
      <!---->
    </li>
  </ul>
</div>
`;

exports[`Storyshots components|m-repeater With items (Custom item template) 1`] = `
<div
  class="m-repeater"
>
  <ul
    class="m-repeater__list"
  >
    <li
      class="m-repeater__row"
    >
      <div
        class="m-repeater__col"
      >
        
                    Item # 0 uuid
                
      </div>
       
      <button
        class="m-icon-button m-repeater__delete m--is-skin-light m--has-ripple"
        style="width: 44px; height: 44px;"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="m-icon m-icon-button__icon"
          height="20px"
          width="20px"
        >
          <!---->
           
          <use
            aria-hidden="true"
            class=""
          />
        </svg>
         
        <!---->
      </button>
    </li>
    <li
      class="m-repeater__row"
    >
      <div
        class="m-repeater__col"
      >
        
                    Item # 1 uuid
                
      </div>
       
      <button
        class="m-icon-button m-repeater__delete m--is-skin-light m--has-ripple"
        style="width: 44px; height: 44px;"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="m-icon m-icon-button__icon"
          height="20px"
          width="20px"
        >
          <!---->
           
          <use
            aria-hidden="true"
            class=""
          />
        </svg>
         
        <!---->
      </button>
    </li>
    <li
      class="m-repeater__footer"
    >
      <a
        class="m-link m-add m--is-unvisited m--is-button m--has-left-icon"
        href="#"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="m-icon m-link__icon"
          height="24px"
          width="24px"
        >
          <!---->
           
          <use
            aria-hidden="true"
            class=""
          />
        </svg>
         
        <span
          class="m-link__text"
        >
          Add
        </span>
         
        <!---->
      </a>
    </li>
  </ul>
</div>
`;

exports[`Storyshots components|m-repeater With items (Custom row template) 1`] = `
<div
  class="m-repeater"
>
  <ul
    class="m-repeater__list"
  >
    <li>
      <button>
        Delete
      </button>
      
                    Item # 0 uuid
                
    </li>
    <li>
      <button>
        Delete
      </button>
      
                    Item # 1 uuid
                
    </li>
    <li
      class="m-repeater__footer m--is-slot-row"
    >
      <a
        class="m-link m-add m--is-unvisited m--is-button m--has-left-icon"
        href="#"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="m-icon m-link__icon"
          height="24px"
          width="24px"
        >
          <!---->
           
          <use
            aria-hidden="true"
            class=""
          />
        </svg>
         
        <span
          class="m-link__text"
        >
          Add
        </span>
         
        <!---->
      </a>
    </li>
  </ul>
</div>
`;
